Smt/Solver.lean

```diff
@@ -40,84 +42,186 @@ register_option smt.solver.path : String := {
   descr := "The path to the solver used for solving the SMT query."
 }
 
-/-- The data-structure for SMT-LIB based solver. -/
-structure Solver where
-  commands : List String
-  deriving Inhabited
+/-- Result of an SMT query. -/
+inductive Result where
+  | sat     : Result
+  | unsat   : Result
+  | unknown : Result
+deriving DecidableEq
+
+instance : ToString Result where
+  toString : Result → String
+    | .sat     => "sat"
+    | .unsat   => "unsat"
+    | .unknown => "unknown"
+
+/-- The data-structure for the state of the generic SMT-LIB solver. -/
+structure SolverState where
+  commands : List Command
+  proc : IO.Process.Child ⟨.piped, .piped, .piped⟩
+
+variable [Monad m] [MonadLiftT IO m]
+
+abbrev SolverT (m) [Monad m] [MonadLiftT IO m] := StateT SolverState m
+
+abbrev SolverM := SolverT IO
 
 namespace Solver
-open Term
 
-variable (solver : Solver)
+private def addCommand (c : Command) : SolverT m Unit := do
+  let state ← get
+  set { state with commands := c :: state.commands }
+  state.proc.stdin.putStr s!"{c}\n"
+  state.proc.stdin.flush
+
+private def getSexp : SolverT m Sexp := do
+  let state ← get
+  let mut out ← state.proc.stdout.getLine
+  let mut sexp := Sexp.parse out
+  while ¬sexp.toBool do
+    out := out ++ (← state.proc.stdout.getLine)
+    sexp := Sexp.parse out
+  if let .ok [.expr [.atom "error", .atom e]] := sexp then
```

Comment:
The DSL should also work in match statements, something like
```suggestion
  if let .ok sexp!{(error {e})} := sexp then
```

Reply:
Good point!